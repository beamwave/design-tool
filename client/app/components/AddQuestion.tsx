import React, { Component } from 'react'
import { connect } from 'react-redux'
import { Link } from 'react-router-dom'
import fontawesome from '@fortawesome/fontawesome'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import Dropzone from 'react-dropzone'
import {
  startBatchCreateQuestions,
  startCreateNewQuestion
} from '../actions/create'
import { editStoreQuiz } from '../actions/quizzes'

interface IProps {
  history?: any
  username?: string
  quizID?: string
  quizUUID?: string
  quizzes?: any[]
  editFunc?: () => any
  edited?: boolean
  clickedAddQuestion?: boolean
  adding?: (any) => any
  quiz?: any
  page?: 1
  editStoreQuiz?: (any) => any
  startCreateNewQuestion?: (any) => any
  startBatchCreateQuestions?: (any) => any
  enableSaveAllButton?: (any) => any
}
// interface IProps extends RouteComponentProps<any> {
//   startCreateNewQuestion: (any) => any
// }
interface AnswerFormat {
  answer: string
  percentPoints: number
  feedback: string
}

interface QuestionFormat {
  author: string
  title: string
  tags: string
  // image?: string;
  answers: AnswerFormat[]
}

export class AddQuestion extends Component<IProps, any> {
  state = {
    errMsg: '',
    displayPhoto: '',
    newQuestions: [],
    currentQuestion: {
      author: this.props.username,
      title: '',
      tags: '',
      answers: [
        {
          answer: '',
          percentPoints: 0,
          feedback: 'no feedback for this question'
        },
        {
          answer: '',
          percentPoints: 0,
          feedback: 'no feedback for this question'
        }
      ],
      format: 'true-false'
    },
    mainView: {
      display: 'inputDisplay',
      questionToDisplay: {
        title: '',
        tags: '',
        answers: []
      }
    },
    editQuestion: 0,
    page: 1,
    edited: this.props.edited,
    clickedAddQuestion: this.props.clickedAddQuestion
  }

  page1 = () => {
    this.setState(prevState => ({ page: 1 }))
    console.log('after page update!!!!!!')
  }

  page2 = () => this.setState(prevState => ({ page: 2 }))

  private updateTitle = (e: any) => {
    const title = e.target.value
    this.setState({
      currentQuestion: {
        ...this.state.currentQuestion,
        title
      }
    })
    this.props.editFunc()
  }

  updateReducerStore = async () => {
    let sendQuestionList = []
    for (let item of this.state.newQuestions) {
      //splits array, trims spaces, and puts the return array into a new set to get rid of duplicates
      let set = new Set(
        item.tags.split(',').map(string => {
          return string.trim()
        })
      )
      //spreads set to convert it into an array
      let testArr: any[] = [...set]
      sendQuestionList.push({
        ...item,
        tags: testArr
      })
    }
    const data = {
      quizID: this.props.quizID,
      newQuestions: sendQuestionList
    }
    let testvar = await this.props.startBatchCreateQuestions(data)
    console.log('updateReducerStore Test Var', testvar)
    if (Array.isArray(testvar)) {
      this.updateStore(sendQuestionList)
      this.setState({
        ...this.state,
        newQuestions: []
      })
      console.log(window.location.href.split('/')[3])
      if (window.location.href.split('/')[3] === 'add-question') {
        this.props.history.push('/quizzes/created')
      }
    } else {
      this.setState({
        ...this.state,
        errMsg: 'Invalid question format. Please edit and try again.'
      })
    }
  }

  private updateFormat = (e: any) => {
    const format = e.target.value
    if (format === 'true-false') {
      this.setState({
        currentQuestion: {
          ...this.state.currentQuestion,
          format: format,
          answers: [
            {
              answer: '',
              percentPoints: 0,
              feedback: 'no feedback for this question'
            },
            {
              answer: '',
              percentPoints: 0,
              feedback: 'no feedback for this question'
            }
          ]
        }
      })
    } else {
      this.setState({
        currentQuestion: {
          ...this.state.currentQuestion,
          format: format,
          answers: [
            {
              answer: '',
              percentPoints: 0,
              feedback: 'no feedback for this question'
            },
            {
              answer: '',
              percentPoints: 0,
              feedback: 'no feedback for this question'
            },
            {
              answer: '',
              percentPoints: 0,
              feedback: 'no feedback for this question'
            },
            {
              answer: '',
              percentPoints: 0,
              feedback: 'no feedback for this question'
            }
          ]
        }
      })
    }
  }

  private addToList = (e: any) => {
    let testValidator = this.formatValidator(this.state.currentQuestion)
    console.log(testValidator)
    if (!testValidator) {
      this.setState({
        ...this.state,
        errMsg:
          'Please make sure your question awards ~100 points and all boxes are filled in.'
      })
    }
    // if (this.state.currentQuestion.title)
    else {
      let questionArr = this.state.newQuestions
      questionArr.push(this.state.currentQuestion)
      this.setState(
        {
          ...this.state,
          newQuestions: questionArr,
          errMsg: '',
          currentQuestion: {
            author: this.props.username,
            title: '',
            tags: '',
            answers: [
              {
                answer: '',
                percentPoints: 0,
                feedback: 'no feedback for this question'
              },
              {
                answer: '',
                percentPoints: 0,
                feedback: 'no feedback for this question'
              }
            ],
            format: 'true-false'
          }
        },
        () => {
          console.log('after answer update! Now changing page')
          this.page1()
        }
      )
      // this.props.enableSaveAllButton(this.state.newQuestions.length)
    } // else...
  }

  private updateArr = (e: any, arg1: number, arg2: string) => {
    let newAnswersArr = this.state.currentQuestion.answers
    newAnswersArr[arg1][arg2] = e.target.value
    this.setState({
      currentQuestion: {
        ...this.state.currentQuestion,
        answers: newAnswersArr
      }
    })
  }

  private updateStore = questionList => {
    if (this.props.quizzes !== [] && this.props.quiz) {
      let modQuiz = this.props.quiz
      modQuiz.questions = modQuiz.questions.concat(questionList)

      let quizList = this.props.quizzes
      for (let i = 0; i < quizList.length; i++) {
        if (quizList[i] !== null && quizList[i].uuid === this.props.quizID) {
          quizList.splice(i, 1, modQuiz)
        }
      }
      console.log('this is the quiz passed in props', this.props.quiz)
      console.log('this is the modQuiz', modQuiz)
      console.log('this is the quizID', this.props.quizID)
      console.log('modified quizList', quizList)
      this.props.editStoreQuiz(quizList)
    }
  }

  // private sendOneQuestion = (e: any) => {
  //   this.props.startCreateNewQuestion(this.state.currentQuestion);
  // };

  // @ts-ignore
  // componentWillReceiveProps = (props, nextProps) =>
  //   console.log("props: ", props, "nextProps: ", nextProps);

  private changeView = (
    typeOfView: string,
    questionToView?: {},
    i?: number
  ) => {
    if (typeOfView === 'questionDisplay') {
      this.setState({
        mainView: {
          display: 'questionDisplay',
          questionToDisplay: questionToView
        },
        editQuestion: i
      })
    } else {
      this.setState({
        mainView: {
          display: 'inputDisplay',
          questionToDisplay: {}
        },
        editQuestion: 0
      })
    }
    console.log('changeViewCalled', this.state)
  }

  private formatValidator = (questionToValidate: QuestionFormat) => {
    let validObject = true
    let percentPoints = 0

    if (!questionToValidate.title || !questionToValidate.author) {
      return (validObject = false)
    } else {
      for (let item of questionToValidate.answers) {
        if (!item.answer) {
          return (validObject = false)
        }
        percentPoints = percentPoints + Number(item.percentPoints)
        if (percentPoints > 101) {
          return false
        }
      }
    }
    console.log('Inside Validator Function', percentPoints, validObject)
    return validObject
    // return percentPoints >= 99 ? true : false
  }

  private printQuestionsArr = () => {
    return (
      <div>
        {this.state.newQuestions.map((item, i) => {
          return (
            <div
              key={'questions' + i}
              onClick={event => {
                this.changeView('questionDisplay', item, i)
                this.page2()
              }}
              className="tab"
            >
              {/* Question {i + 1} */}
            </div>
          )
        })}
      </div>
    )
  }

  private onDrop = (files: any) => {
    const file = files[0]
    const getBase64 = async file => {
      var reader = new FileReader()
      reader.readAsDataURL(file)
      let codeString
      reader.onload = function() {
        codeString = reader.result
        codeString = codeString.split(',', 2)
        this.setState({
          ...this.state,
          currentQuestion: {
            ...this.state.currentQuestion,
            image: codeString[1]
          },
          displayPhoto: codeString.join(',')
        })
      }.bind(this)
      reader.onerror = function(error) {
        console.log('Error: ', error)
      }
      return codeString
    }
    getBase64(file)
  }

  private editQuestionElements = (e: any, propertyName) => {
    const index = this.state.editQuestion
    let updatedQuestions = this.state.newQuestions
    updatedQuestions[index] = {
      ...updatedQuestions[index],
      [propertyName]: e.target.value
    }
    console.log('propertyname', propertyName)
    this.setState({
      newQuestions: updatedQuestions,
      mainView: {
        ...this.state.mainView,
        questionToDisplay: updatedQuestions[index]
      }
    })
  }

  private editQuizAnswers = (e: any, propertyName, i: number) => {
    const index = this.state.editQuestion
    let updatedQuestions = this.state.newQuestions
    updatedQuestions[index].answers[i] = {
      ...updatedQuestions[index].answers[i],
      [propertyName]: e.target.value
    }
    this.setState({
      newQuestions: updatedQuestions,
      mainView: {
        ...this.state.mainView,
        questionToDisplay: updatedQuestions[index]
      }
    })
  }

  private editTagElement = (e: any) => {
    this.setState({
      currentQuestion: {
        ...this.state.currentQuestion,
        tags: e.target.value
      }
    })
  }

  private renderMainDisplayElement = () => {
    switch (this.state.mainView.display) {
      case 'questionDisplay':
        return (
          <div className="details">
            <div className="group">
              <label htmlFor="editQuestionTitle">Title:</label>
              <input
                type="text"
                id="editQuestionTitle"
                value={this.state.mainView.questionToDisplay.title}
                onChange={e => {
                  this.editQuestionElements(e, 'title')
                }}
              />
            </div>
            <div className="group">
              <label htmlFor="editQuestionTag">Tags:</label>
              <input
                type="text"
                id="editQuestionTag"
                value={this.state.mainView.questionToDisplay.tags}
                onChange={e => {
                  this.editQuestionElements(e, 'tags')
                }}
              />
            </div>
            {this.state.mainView.questionToDisplay.answers.map((item, i) => {
              return (
                <div className="question-fields">
                  <div>
                    <div className="group">
                      <label htmlFor="true-false-answer">Answer</label>
                      <input
                        type="text"
                        id="true-false-answer"
                        value={item.answer}
                        onChange={e => {
                          this.editQuizAnswers(e, 'answer', i)
                        }}
                      />
                    </div>
                    <div className="group">
                      <label htmlFor="true-false-percent-points">Score</label>
                      <input
                        type="text"
                        id="true-false-percent-points"
                        value={item.percentPoints}
                        onChange={e => {
                          this.editQuizAnswers(e, 'percentPoints', i)
                        }}
                      />
                    </div>
                    {/* <div className="group">
                    <label htmlFor="true-false-feed-back">feed Back</label>
                    <input
                      type="text"
                      id="true-false-feed-back"
                      value={item.feedback}
                      onChange={e => {
                        this.editQuizAnswers(e, 'feedback', i)
                      }}
                    />
                  </div> */}
                  </div>
                </div>
              )
            })}
            <button
              onClick={() => {
                this.changeView('inputDisplay')
                this.page1()
              }}
              className="back-to-create-question"
            >
              <FontAwesomeIcon icon="arrow-left" className="icon" />
              {/* Create New Question */}
            </button>
          </div>
        )
      default:
        return <div className="question-fields">{this.createQuestions()}</div>
    }
  }

  private createQuestions = () => {
    switch (this.state.currentQuestion.format) {
      case 'true-false':
        return (
          <div>
            <div className="group">
              <label htmlFor="true-false-answer">Answer 1</label>
              <input
                type="text"
                id="true-false-answer"
                value={this.state.currentQuestion.answers[0].answer}
                onChange={(e: any) => {
                  this.updateArr(e, 0, 'answer')
                }}
              />
            </div>
            <div className="group">
              <label htmlFor="true-false-percent-points">Score</label>
              <input
                type="text"
                id="true-false-percent-points"
                value={this.state.currentQuestion.answers[0].percentPoints}
                onChange={(e: any) => {
                  this.updateArr(e, 0, 'percentPoints')
                }}
              />
            </div>
            {/* <div className="group">
              <label htmlFor="true-false-feed-back">feed Back</label>
              <input
                type="text"
                id="true-false-feed-back"
                value={this.state.currentQuestion.answers[0].feedback}
                onChange={(e: any) => {
                  this.updateArr(e, 0, 'feedback')
                }}
              /> 
            </div> */}
            <div className="group">
              <label htmlFor="true-false-answer">Answer 2</label>
              <input
                type="text"
                id="true-false-answer"
                value={this.state.currentQuestion.answers[1].answer}
                onChange={(e: any) => {
                  this.updateArr(e, 1, 'answer')
                }}
              />
            </div>
            <div className="group">
              <label htmlFor="true-false-percent-points">Score</label>
              <input
                type="text"
                id="true-false-percent-points"
                value={this.state.currentQuestion.answers[1].percentPoints}
                onChange={(e: any) => {
                  this.updateArr(e, 1, 'percentPoints')
                }}
              />
            </div>
            {/* <div className="group">
              <label htmlFor="true-false-feed-back">feed Back</label>
              <input
                type="text"
                id="true-false-feed-back"
                value={this.state.currentQuestion.answers[1].feedback}
                onChange={(e: any) => {
                  this.updateArr(e, 1, 'feedback')
                }}
              />
            </div> */}

            {/* <button type="button" onClick={this.addToList}>
              Add Question
            </button> */}
          </div>
        )
      default:
        return (
          <div>
            <div className="group">
              <label htmlFor="true-false-answer">Answer</label>
              <input
                type="text"
                id="true-false-answer"
                value={this.state.currentQuestion.answers[0].answer}
                onChange={(e: any) => {
                  this.updateArr(e, 0, 'answer')
                }}
              />
            </div>
            <div className="group">
              <label htmlFor="true-false-percent-points">Score</label>
              <input
                type="text"
                id="true-false-percent-points"
                value={this.state.currentQuestion.answers[0].percentPoints}
                onChange={(e: any) => {
                  this.updateArr(e, 0, 'percentPoints')
                }}
              />
            </div>
            {/* <div className="group">
              <label htmlFor="true-false-feed-back">feed Back</label>
              <input
                type="text"
                id="true-false-feed-back"
                value={this.state.currentQuestion.answers[0].feedback}
                onChange={(e: any) => {
                  this.updateArr(e, 0, 'feedback')
                }}
              />
            </div> */}
            <div className="group">
              <label htmlFor="true-false-answer">Answer</label>
              <input
                type="text"
                id="true-false-answer"
                value={this.state.currentQuestion.answers[1].answer}
                onChange={(e: any) => {
                  this.updateArr(e, 1, 'answer')
                }}
              />
            </div>
            <div className="group">
              <label htmlFor="true-false-percent-points">Score</label>
              <input
                type="text"
                id="true-false-percent-points"
                value={this.state.currentQuestion.answers[1].percentPoints}
                onChange={(e: any) => {
                  this.updateArr(e, 1, 'percentPoints')
                }}
              />
            </div>
            {/* <div className="group">
              <label htmlFor="true-false-feed-back">feed Back</label>
              <input
                type="text"
                id="true-false-feed-back"
                value={this.state.currentQuestion.answers[1].feedback}
                onChange={(e: any) => {
                  this.updateArr(e, 1, 'feedback')
                }}
              />
            </div> */}
            <div className="group">
              <label htmlFor="true-false-answer">Answer 1</label>
              <input
                type="text"
                id="true-false-answer"
                value={this.state.currentQuestion.answers[2].answer}
                onChange={(e: any) => {
                  this.updateArr(e, 2, 'answer')
                }}
              />
            </div>
            <div className="group">
              <label htmlFor="true-false-percent-points">Score</label>
              <input
                type="text"
                id="true-false-percent-points"
                value={this.state.currentQuestion.answers[2].percentPoints}
                onChange={(e: any) => {
                  this.updateArr(e, 2, 'percentPoints')
                }}
              />
            </div>
            {/* <div className="group">
              <label htmlFor="true-false-feed-back">feed Back</label>
              <input
                type="text"
                id="true-false-feed-back"
                value={this.state.currentQuestion.answers[2].feedback}
                onChange={(e: any) => {
                  this.updateArr(e, 2, 'feedback')
                }}
              />
            </div> */}
            <div className="group">
              <label htmlFor="true-false-answer">Answer 2</label>
              <input
                type="text"
                id="true-false-answer"
                value={this.state.currentQuestion.answers[3].answer}
                onChange={(e: any) => {
                  this.updateArr(e, 3, 'answer')
                }}
              />
            </div>
            <div className="group">
              <label htmlFor="true-false-percent-points">Score</label>
              <input
                type="text"
                id="true-false-percent-points"
                value={this.state.currentQuestion.answers[3].percentPoints}
                onChange={(e: any) => {
                  this.updateArr(e, 3, 'percentPoints')
                }}
              />
            </div>
            {/* <div className="group">
               <label htmlFor="true-false-feed-back">feed Back</label>
               <input
                 type="text"
                 id="true-false-feed-back"
                 value={this.state.currentQuestion.answers[3].feedback}
                 onChange={(e: any) => {
                   this.updateArr(e, 3, 'feedback')
                 }}
               />
             </div> */}
            {/* <button type="button" onClick={this.addToList}>
              Add Question
            </button> */}
          </div>
        )
    }
  }

  // @ts-ignore
  render = () => {
    const { startCreateNewQuestion } = this.props
    const { page, displayPhoto } = this.state
    // const {
    //   currentQuestion: { format },
    //   mainView: { display }
    // } = this.state
    // /user/updatePoints
    // post
    return (
      <div className="add-question-container">
        <p className="title">Add Question</p>
        <div className="close" onClick={this.props.adding}>
          <FontAwesomeIcon icon="times" />
        </div>
        <div className="que-question-button" onClick={this.addToList}>
          <FontAwesomeIcon icon="plus" />
        </div>
        {page === 1 && (
          <div className="details">
            <div className="group">
              <label htmlFor="title">Question</label>
              <textarea
                id="title"
                value={this.state.currentQuestion.title}
                onChange={this.updateTitle}
                rows={3}
                data-enable-grammarly="false"
              />
            </div>
            <div className="split-group">
              <div className="group">
                <label htmlFor="tags">Tags</label>
                <input
                  type="text"
                  id="tags"
                  value={this.state.currentQuestion.tags}
                  onChange={this.editTagElement}
                />
              </div>
              <div className="group">
                <label htmlFor="type">Type</label>
                <FontAwesomeIcon className="icon" icon="angle-down" />
                <select
                  name="type"
                  id="question-options-dropdown"
                  value={this.state.currentQuestion.format}
                  onChange={this.updateFormat}
                >
                  <option value="true-false">True-or-False</option>
                  <option value="multiple-choice">Multiple Choice</option>
                  <option value="multiple-select">Multiple Select</option>
                </select>
              </div>
            </div>
            <div className="group">
              <label>Image</label>
              <Dropzone
                onDrop={this.onDrop}
                className={displayPhoto ? 'dropzone' : 'dropzone with-dots'}
              >
                <p className={displayPhoto ? 'white-text' : ''}>+ Upload</p>
                {displayPhoto && <img src={displayPhoto} alt="img" />}
              </Dropzone>
            </div>
          </div>
        )}
        {page === 2 && this.renderMainDisplayElement()}
        {this.state.errMsg && this.state.errMsg}
        <div className="page-navigation">
          <p
            className={page === 1 ? 'page-1 active' : 'page-1'}
            onClick={this.page1}
          >
            1
          </p>
          <p
            className={page === 2 ? 'page-2 active' : 'page-2'}
            onClick={this.page2}
          >
            2
          </p>
        </div>
        {page === 2 && (
          <button
            disabled={
              !this.props.edited || this.state.newQuestions.length === 0
                ? true
                : false
            }
            onClick={this.updateReducerStore}
            className="save-question"
          >
            Save all questions
          </button>
        )}
        <div className="tabs">{this.printQuestionsArr()}</div>
      </div>
    )
  }
}

const mapStateToProps = (state, parentProps) => ({
  username: state.auth.username,
  quizID: parentProps.quizID || state.create.quizID,
  quizzes: state.quizzes.all,
  quiz:
    state.quizzes.all !== [] &&
    state.quizzes.all.find(
      quiz =>
        quiz.uuid === parentProps.quizID || quiz.uuid === state.create.quizID
    )
})

export default connect<any, IProps>(
  mapStateToProps,
  { startBatchCreateQuestions, startCreateNewQuestion, editStoreQuiz }
)(AddQuestion)
