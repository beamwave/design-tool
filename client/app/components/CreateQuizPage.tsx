import React, { Component } from 'react'
import { connect } from 'react-redux'
import { Link } from 'react-router-dom'
import fontawesome from '@fortawesome/fontawesome'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { startCreateNewQuiz } from '../actions/create'

interface IProps {
  history?: any
  username?: any
  quizzes?: any
  startCreateNewQuiz: (quiz: any, any) => any
}

export class CreateQuizPage extends Component<IProps, any> {
  state = {
    quiz: {
      author: this.props.username,
      title: '',
      description: '',
      isPrivate: true,
      type: 'quiz'
    },
    quizzes: this.props.quizzes
  }

  private updateQuiz = (e: any, arg1: string) => {
    switch (arg1) {
      case 'title':
        this.setState({
          quiz: {
            ...this.state.quiz,
            title: e.target.value
          }
        })
        break
      case 'description':
        this.setState({
          quiz: {
            ...this.state.quiz,
            description: e.target.value
          }
        })
      default:
        break
    }
    // console.log(this.state)
  }

  private updatePrivate = (e: any) => {
    switch (e.target.value) {
      case 'private':
        this.setState({
          quiz: {
            ...this.state.quiz,
            isPrivate: true
          }
        })
        break

      default:
        this.setState({
          quiz: {
            ...this.state.quiz,
            isPrivate: false
          }
        })
        break
    }
  }

  private submitQuiz = async (e: any) => {
    e.preventDefault()
    const { startCreateNewQuiz, username } = this.props
    const { quiz } = this.state

    // save quiz to db and update quizzes
    const createAction = await startCreateNewQuiz(quiz, this.props.history)

    // this.props.history.push('/quizzes/created')
    // this.props.history.push('/edit-quiz/' + uuid)
    // .then(res => {
    //   console.log('add quiz props')
    //   console.log(this.props)
    //   this.props.history.push('/add-question')
    // })
    // .catch(err =>
    //   console.log(
    //     'make sure the user knows that the quiz did not make it through'
    //   )
    // )
  }

  // @ts-ignore
  render = () => {
    const {
      quiz: { title }
    } = this.state
    return (
      <div className="create-quiz-page">
        <main>
          <div
            className={
              title.length > 0
                ? 'create-quiz-container valid'
                : 'create-quiz-container invalid'
            }
          >
            <p className="title">Create a quiz</p>
            <form className="details">
              <div className="group">
                <label htmlFor="create-quiz-title">Title</label>
                <input
                  type="text"
                  id="create-quiz-title"
                  value={this.state.quiz.title}
                  onChange={(e: any) => {
                    this.updateQuiz(e, 'title')
                  }}
                />
              </div>
              <div className="group">
                <label htmlFor="create-quiz-description">Description</label>
                <textarea
                  id="create-quiz-description"
                  value={this.state.quiz.description}
                  rows={3}
                  onChange={(e: any) => {
                    this.updateQuiz(e, 'description')
                  }}
                />
              </div>
              <div className="group">
                <label htmlFor="owner">Owner</label>
                <FontAwesomeIcon className="icon" icon="angle-down" />
                <select
                  name="owner"
                  id="question-options-dropdown"
                  onChange={this.updatePrivate}
                >
                  <option value="choose" hidden>
                    select
                  </option>
                  <option value="private">private</option>
                  <option value="public">public</option>
                </select>
              </div>
              <button
                type="button"
                className="save-quiz"
                onClick={this.submitQuiz}
                disabled={title.length > 0 ? false : true}
              >
                Create
              </button>
            </form>
          </div>
        </main>
      </div>
    )
  }
}

const mapStateToProps = state => ({
  username: state.auth.username,
  quizzes: state.quizzes.all
})

export default connect<any, IProps>(
  mapStateToProps,
  { startCreateNewQuiz }
)(CreateQuizPage)
