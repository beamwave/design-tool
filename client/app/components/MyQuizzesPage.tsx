import React, { Component } from 'react'
import moment from 'moment'
import numeral from 'numeral'
import { connect } from 'react-redux'
import { Link } from 'react-router-dom'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { clearResults, clearQuizAttempt } from '../actions/quizzes'
import { loadModal } from '../actions/modal'
import { DELETE_QUIZ_MODAL } from '../constants/modaltypes'
import Pagination from './Pagination'

interface IProps {
  username: string
  quizzes: any[]
  type: string
  quizAttempts: any[]
  clearResults?: () => any
  clearQuizAttempt?: (reset: number) => void
  loadModal?: (type: string, title: string, uuid: any) => any
}

const colors = [
  '#f26161',
  '#cc5252',
  '#e573b1',
  '#ff9d66',
  '#2ddbbc',
  '#42a6a6',
  '#2d9cdb',
  '#3f388c',
  '#4775b2',
  '#b866cc',
  '#333ea6',
  '#f2c94c'
]

export class MyQuizzesPage extends Component<IProps, any> {
  constructor(props) {
    super(props)
    this.state = {
      quizItems: this.props.quizzes,
      attemptItems: this.props.quizAttempts,
      pageOfItems: []
    }
  }
  params = window.location.href.split('/')
  pageType = this.params[4]

  onChangePage = pageOfItems => {
    // update state with new page of items
    this.setState({ pageOfItems: pageOfItems })
  }

  clearResults = (e: any) => {
    this.props.clearResults()
    this.props.clearQuizAttempt(0)
    console.log('clearing results')
  }

  deleteQuizModal = data => (e: any) =>
    this.props.loadModal(DELETE_QUIZ_MODAL, data, null)

  deleteQuestion = e => console.log('quiz uuid: ', e.target.dataset.uuid)

  //@ts-ignore
  render = () => {
    const { quizzes, type, quizAttempts, username } = this.props
    const { pageOfItems, quizItems, attemptItems } = this.state
    return (
      <div className="my-quizzes-page">
        {type === 'created' && (
          <p className="links">
            <Link to="/quizzes/created" className="quiz-type">
              created
            </Link>&nbsp;/&nbsp;<Link
              to="/quizzes/taken"
              onClick={this.clearResults}
            >
              taken
            </Link>
          </p>
        )}
        {type === 'taken' && (
          <p className="links">
            <Link to="/quizzes/created">created</Link>&nbsp;/&nbsp;<Link
              to="/quizzes/taken"
              className="quiz-type"
              onClick={this.clearResults}
            >
              taken
            </Link>
          </p>
        )}
        {this.state.pageOfItems.map(item => (
          <div key={item.id}>{item.name}</div>
        ))}
        <div className="blocks">
          {this.pageType === 'created' &&
            quizzes !== undefined &&
            quizzes.map((quiz: any) => {
              // pageOfItems.map((quiz: any) => {
              if (quiz !== null && quiz.author === username)
                return (
                  <div key={quiz.uuid} className="block">
                    <div
                      className="delete-quiz"
                      data-uuid={quiz.uuid}
                      onClick={this.deleteQuizModal(quiz.title)}
                    >
                      <FontAwesomeIcon icon="trash" />
                      <p className="hint">Permanently delete this quiz</p>
                    </div>
                    <Link className="quiz-link" to={`/view-quiz/${quiz.uuid}`}>
                      <h1 className="name">{quiz.title}</h1>
                      <p className="amount">
                        {quiz.questions.length} questions
                      </p>
                      <div className="tags">
                        {quiz.tags.length === 0 && (
                          <p className="tag-null-set">No tags</p>
                        )}
                        {quiz.tags.length > 0 && quiz.tags.length > 3
                          ? quiz.tags.slice(0, 3).map(tag => (
                              <div key={tag.allLowerCase} className="tag">
                                <div
                                  className="tag-dot"
                                  style={{
                                    height: 8,
                                    width: 8,
                                    borderRadius: 50,
                                    backgroundColor:
                                      colors[
                                        Math.floor(
                                          Math.random() * colors.length
                                        )
                                      ]
                                  }}
                                />
                                <p className="tag-text shorten">
                                  {tag.allLowerCase}
                                </p>
                              </div>
                            ))
                          : quiz.tags.map(tag => (
                              <div key={tag.allLowerCase} className="tag">
                                <div
                                  className="tag-dot"
                                  style={{
                                    height: 8,
                                    width: 8,
                                    borderRadius: 50,
                                    backgroundColor:
                                      colors[
                                        Math.floor(
                                          Math.random() * colors.length
                                        )
                                      ]
                                  }}
                                />
                                <p className="tag-text full">
                                  {tag.allLowerCase}
                                </p>
                              </div>
                            ))}
                        {quiz.tags.length > 3 && (
                          <p className="extra-tags">+{quiz.tags.length - 3}</p>
                        )}
                      </div>
                      <div className="fill-block"></div>
                      {/* {quiz.tags.map(tag => (
                    <div key={tag.allLowerCase} className="tag">
                      {tag.allLowerCase}
                    </div>
                  ))} */}
                    </Link>
                  </div>
                )
            })}
          {this.pageType === 'taken' &&
            quizAttempts !== undefined &&
            quizAttempts.length > 0 &&
            // quizAttempts.length > 0 &&
            quizAttempts.map(
              (quizAttempt: any, index) =>
                quizAttempt.timings.finished !== undefined && (
                  <Link
                    to={`/review-quiz/${quizAttempt.quizUUID}/${index}`}
                    key={index}
                  >
                    <div className="block">
                      <h1 className="name">{quizAttempt.title}</h1>
                      {quizAttempt.questions !== undefined && (
                        <p className="amount">
                          {quizAttempt.questions.length} questions
                        </p>
                      )}
                      <p className="score">Your Score: {quizAttempt.score}</p>
                    </div>
                  </Link>
                )
            )}
          {this.pageType === 'taken' &&
            quizAttempts !== undefined &&
            (quizAttempts.every(
              quizAttempt => quizAttempt.timings.finished === undefined
            ) ||
              quizAttempts.length === 0) && (
              <p className="no-taken-quizzes">
                You have not taken any quizzes yet.
              </p>
            )}
        </div>
        <div>
          {/* {this.pageType === 'created' &&
            quizzes !== undefined &&
            quizzes.length > 6 &&
            quizzes.map((quiz, i) => {
              return <p className="page-number">{Math.ceil(i / 6)}</p>
            })} */}
          <Pagination
            items={type === 'taken' ? quizItems : attemptItems}
            onChangePage={this.onChangePage}
            initialPage={1}
          />
        </div>
      </div>
    )
  }
}

const mapStateToProps = (state, props) => ({
  username: state.auth.username,
  quizzes: state.quizzes.all,
  quizAttempts: state.quizzes.allAttempts,
  type: props.match.params.type
})

export default connect(
  mapStateToProps,
  {
    clearResults,
    loadModal,
    clearQuizAttempt
  }
)(MyQuizzesPage)
