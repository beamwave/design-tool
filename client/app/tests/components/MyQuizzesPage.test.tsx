import React, { Component } from 'react'
import { shallow } from 'enzyme'
import { MyQuizzesPage } from '../../components/MyQuizzesPage'

/* add mock data for quizzes and quizAttempts (rename to allAttempts or vice versa) */

test('should render MyQuizzesPage component', () => {
  const wrapper = shallow(
    <MyQuizzesPage
      username="PappalooseGoose"
      type="placeholder"
      quizzes={['quizzes array placeholder']}
      quizAttempts={['quizAttempts array placeholder']}
    />
  )
  expect(wrapper).toMatchSnapshot()
})
