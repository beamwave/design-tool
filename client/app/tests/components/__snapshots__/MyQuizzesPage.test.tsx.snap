// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render MyQuizzesPage component 1`] = `
<div
  className="my-quizzes-page"
>
  <div
    className="blocks"
  />
  <div>
    <Pagination
      initialPage={1}
      items={
        Array [
          "quizAttempts array placeholder",
        ]
      }
      onChangePage={[Function]}
    />
  </div>
</div>
`;
